lib/concurrent/timer_task2.rb

```diff
@@ -0,0 +1,297 @@
+    #   @option opts [Integer] :timeout_interval number of seconds a task can
+    #     run before it is considered to have failed (default: TIMEOUT_INTERVAL)
+    #   @option opts [Boolean] :run_now Whether to run the task immediately
+    #     upon instantiation or to wait until the first execution_interval
+    #     has passed (default: false)
+    #   @option opts [Symbol] :reschedule when to schedule next execution relative,
+    #     relative to a current one, `:before` or `:after` (default: :after)
+    #   @option opts [Promises::Channel, nil] :channel if given, execution results
+    #     will be pushed into the channel (default: nil)
+    #
+    #   @raise ArgumentError when no block is given.
+    #
+    #   @yield to the block after :execution_interval seconds have passed since
+    #     the last yield
+    #   @yieldparam task a reference to the `TimerTask` instance so that the
+    #     block can control its own lifecycle. Necessary since `self` will
+    #     refer to the execution context of the block rather than the running
+    #     `TimerTask`.
+    #   @yieldparam cancellation a cancellation object representing the joined cancellation
+    #     of the timer task and this run's timeout
+    #
+    #   @return [TimerTask] the new `TimerTask`
+    def initialize(opts = {}, &task)
+      raise ArgumentError.new('no block given') unless block_given?
+      raise ArgumentError.new('reschedule must be either :before or :after') unless [nil, :before, :after].include?(opts[:reschedule])
+      @execution_interval = AtomicReference.new nil
+      @timeout_interval = AtomicReference.new nil
+      @cancellation = AtomicReference.new nil
+
+      self.execution_interval = opts[:execution] || opts[:execution_interval] || EXECUTION_INTERVAL
+      self.timeout_interval = opts[:timeout] || opts[:timeout_interval] || TIMEOUT_INTERVAL
+
+      @Channel = opts[:channel]
+
+      @Executor = SafeTaskExecutor.new(task)
+      @Reschedule = opts[:reschedule] || :after
+      @RunNow = opts[:now] || opts[:run_now]
+    end
+
+    # Create and execute a new `TimerTask2`.
+    #
+    # @!macro timer_task_initializ
+    #
+    # @example
+    #   task = Concurrent::TimerTask2.execute(execution_interval: 10) { print "Hello World\n" }
+    #   task.running? #=> true
+    def self.execute(*args, &task)
+      self.new(*args, &task).execute
+    end
+
+    # Execute a previously created `TimerTask2`.
+    #
+    # @return [TimerTask2] a reference to `self`
+    #
+    # @example Instance and execute in separate steps
+    #   task = Concurrent::TimerTask2.new(execution_interval: 10) { print "Hello World\n" }
+    #   task.running? #=> false
+    #   task.execute
+    #   task.running? #=> true
+    #
+    # @example Instance and execute in one line
+    #   task = Concurrent::TimerTask2.new(execution_interval: 10) { print "Hello World\n" }.execute
+    #   task.running? #=> true
+    def execute
+      return if running?
```

Comment:
should return self as well

Reply:
However, it would be food to have `execute?` which returns true of false as well.